### DISCUSSION TITLE
Float or double

### DISCUSSION THREAD

**Discussion by charlenni** [Mapsui/Mapsui#2368] [Development]
I found, that we use sometimes float, sometimes double. It isn't very consistent. Perhaps it would be good to use one for all/the most.

Any comments?

**Comment by pauldendulk** [comment]
We have a large code base. There can be reasons to choose a float or double. A float should be faster (and performance matters for Mapsui). Sometimes it matters that it is smaller. A double has more precision. Sometimes this is necessary. For instance in SphericalMercator.From/ToLonLat we converted to float which [caused quite noticeable errors](https://github.com/Mapsui/Mapsui/issues/1212). So we should not make a general choice for float or double.

Perhaps we can make a choice in some specific context. If it is about a position within the canvas a float should be enough. But if you do calculations (like transformations) on that float precision could matter. 

Another thing to take into account it the precision of the components we use. An SKPoint has float x and y, while a NTS Coordinate has a double _x and _y. If we assign those to our own components we usually would like to use the same type.

Where did you run into this?

**Reply by charlenni** [reply]
I always stumble about this. If you work with SkiaSharp and you use double, you have to convert all the double. The last time I saw it while refactoring the widgets. Is Width and Height an integer, a float or a double?

**Reply by pauldendulk** [reply]
With "Where did you run into this?" I meant to ask in which specific situation you ran into this so that we can make a decision about that context. Apparently it is about position in widgets.

I just had a look at MAUI, WinUI and Avalonia and they all use double consistently for things like Width/Height/Rect/Point/Size. So that makes things easy. We should use double everywhere for widgets.

**Comment by pauldendulk** [comment]
The rounding error in SphericalMercator we had was unnecessary because we worked with double everywhere but rounded to float anyway. So, totally unnecessary. This was probably because at some point some of these variables were floats. We will have this problem in more places and need to fix it.

**Comment by pauldendulk** [comment]
If I search the entire code base for **'float' I get 397 hits**. If I search for **'double' I get 677 hits**. If I skim through some of those I get the impression a lot of floats could be doubles.

**Reply by pauldendulk** [reply]
205 of the 'float' hits are in Mapsui.Rendering.Skia. We will need them there because skia uses float. 

So, only 197 'float' hits in the remainder of the code. If I go through those it seems a lot of them should be double and others could be double. 

For instance MapControl.ViewportWidth/Height should be double. The UI uses double and our Viewport uses double but then there is the float field in between.

**Reply by charlenni** [reply]
Ok, then you prefer double for all this. Will change the variables in the Widget to this.

**Comment by pauldendulk** [comment]
Here in the CalloutRenderer we convert the rotation to float early on. We should do that later, at the point where we call the skia api:
https://github.com/Mapsui/Mapsui/blob/d8790714c5329a02c607f3259b543cb91c2095e9/Mapsui.Rendering.Skia/SkiaStyles/CalloutStyleRenderer.cs#L61
